deep_qa/layers/attentive_gru.py

```diff
@@ -1,76 +1,19 @@
-"""
-Knowledge combiners take:
-
-- Encoded representations of background facts related to the sentence
-- Attention weights over the background as a single tensor.
-
-These are then combined in some way to return a single representation of the
-background knowledge per sample. The simplest way for this to happen is simply
-taking a weighted average of the knowledge representations with respect to the
-attention weights.
-
-Input shapes:
-    - (samples, knowledge_len, input_dim + 1)
-
-Output shape:
-    - (samples, input_dim)
-"""
-
-from collections import OrderedDict
 from overrides import overrides
 
 from keras.engine import InputSpec
 from keras import backend as K
-from keras.layers import Layer
 from keras.layers.recurrent import GRU, _time_distributed_dense
 
 
-class WeightedAverageKnowledgeCombiner(Layer):
-    '''
-    A WeightedAverageKnowledgeCombiner takes a tensor formed by prepending an attention mask
-    onto an encoded representation of background knowledge. Here, we simply split off the
-    attention mask and use it to take a weighted average of the background vectors.
-    '''
-
-    def __init__(self, **kwargs):
-        self.input_spec = [InputSpec(ndim=3)]
-        self.name = kwargs.pop('name')
-        # These parameters are passed for consistency with the
-        # AttentiveGRUKnowlegeCombiner. They are not used here.
-        kwargs.pop('output_dim')
-        kwargs.pop('input_length')
-        super(WeightedAverageKnowledgeCombiner, self).__init__(**kwargs)
-
-    @overrides
-    def call(self, inputs):
-        attention = inputs[:, :, 0]  # (samples, knowledge_length)
-        inputs = inputs[:, :, 1:]  # (samples, knowledge_length, word_dim)
-        return K.sum(K.expand_dims(attention, 2) * inputs, 1)
-
-    def compute_output_shape(self, input_shape):
-        return (input_shape[0], input_shape[2] - 1)
-
-    @overrides
-    def get_config(self):
-        config = {
-                "output_dim": -1,
-                "input_length": -1
-                }
-        base_config = super(WeightedAverageKnowledgeCombiner, self).get_config()
-        config.update(base_config)
-        return config
-
-
-class AttentiveGRUKnowledgeCombiner(GRU):
-    '''
+class AttentiveGRU(GRU):
+    """
     GRUs typically operate over sequences of words. Here we are are operating over the background knowledge
```

Comment:
I'd update this to remove reference to "background knowledge" - as you said, this is now just a general GRU that takes an additional input (e.g., this would still work if you ran it over a single sentence).  I'd also move it into the `encoders/` module.